server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,208 @@
+            fetchAttributes(additionalFields, mail),
+            fetchPerRecipientsHeaders(additionalFields, mail),
+            fetchHeaders(additionalFields, mail),
+            fetchBody(additionalFields, mail),
+            fetchMessageSize(additionalFields, mail));
+    }
+
+    private static Optional<Long> fetchMessageSize(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.MESSAGE_SIZE)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(new Long(mail.getMessageSize()));
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.MESSAGE_SIZE, e);
+        }
+    }
+
+    private static Optional<String> fetchBody(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function((Message message) -> extractor.extract(message)).sneakyThrow())
+                    .flatMap(extractBody());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.BODY, e);
+        }
+    }
+
+    private static Function<? super MessageContent, Optional<String>> extractBody() {
+        return (content) -> {
+            Optional<String> body = content.getTextBody();
+            if (body.isPresent()) {
+                return body;
+            } else {
+                return content.getHtmlBody();
+            }
+        };
+    }
+
+    private static Message convertMessage(MimeMessage message) throws IOException, MessagingException {
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        message.writeTo(rawMessage);
+        return Message.Builder
+                .of()
+                .use(MimeConfig.PERMISSIVE)
+                .parse(new ByteArrayInputStream(rawMessage.toByteArray()))
+                .build();
+    }
+
+    private static Optional<HeadersDto> fetchHeaders(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HEADERS)) {
+            return Optional.empty();
+        }
+
+        try {
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function((MimeMessage message) -> {
+                        HeadersDto headers = new HeadersDto();
+                        Collections
+                            .list(message.getAllHeaders())
+                            .forEach((header) -> headers.add(header.getName(), header.getValue()));
```

Comment:
Still some mutability around? Please do use `Steam::collect` with the `Guavate.toImmutableListMultimap()`, you gonna love it!